src/extra/rust_interop.cpp

```diff
@@ -0,0 +1,46 @@
+#include "gfx/shards_types.hpp"
+#include "gfx.hpp"
+#include <foundation.hpp>
+
+using shards::Var;
+namespace gfx {
+SHTypeInfo *getMainWindowGlobalsType() {
+  static SHTypeInfo type = gfx::MainWindowGlobals::Type;
+  return &type;
+}
+const char *getMainWindowGlobalsVarName() { return gfx::Base::mainWindowGlobalsVarName; }
+SHTypeInfo *getQueueType() {
+  static SHTypeInfo type = Types::DrawQueue;
+  return &type;
+}
+
+template <typename T> T *castChecked(const SHVar &var, const shards::Type &type) {
+  SHTypeInfo typeInfo(type);
+  if (var.valueType != SHType::Object)
+    throw std::logic_error("Invalid type");
+  if (var.payload.objectVendorId != typeInfo.object.vendorId)
+    throw std::logic_error("Invalid object vendor id");
+  if (var.payload.objectTypeId != typeInfo.object.typeId)
+    throw std::logic_error("Invalid object type id");
+  return reinterpret_cast<T *>(var.payload.objectValue);
+}
+
+SHVar MainWindowGlobals_getDefaultQueue(const SHVar &mainWindowGlobals) {
```

Comment:
thread safety of all those is dubious though.... maybe they should be thread_local

Reply:
Are you talking about getMainWindowGlobalsType/getQueueType ?

Apparently since C++11 these are considered thread safe https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

Reply:
Not really well defined but anyways do we assume that only one instance of those is ever needed in a single process? Feels like bgfx 😅

Reply:
MainWindowGlobals_getDefaultQueue just gets a value inside this context variable. You could still technically have multiple windows (each has their own `mainWindowGlobals` but is set so eveything inside :Content can access it)

I only use static in this file for the types, which don't change